### PR TITLE
improve: Add tests for v3 SpeedUps

### PR DESCRIPTION
**Pull request by pxrl**
Most of the existing v2 tests are replicated, with some exceptions and modifications. The "early" tests are specifically ignored because v3 isn't anywhere near as sensitive to deposits with a quoteTimestamp that's ahead of the current HubPool timestamp.

### CODE REVIEW

File: test/SpokePoolClient.SpeedUp.ts
Comment (pxrl) [inline test/SpokePoolClient.SpeedUp.ts:1]: Most of the changes in this file are just indenting the existing tests under a "v2" `describe` section. There are no functional changes, but unfortunately the diff doesn't really show that the change is only indentation for the most part.

File: test/SpokePoolClient.SpeedUp.ts [hunk 1]
```diff
@@ -33,264 +31,376 @@ describe("SpokePoolClient: SpeedUp", function () {
...
-      deposit.depositId,
-      deposit.recipient,
-      "0x",
-      speedUpSignature.signature
-    );
-
-    let success = false;
-    try {
+
+  describe("v3", function () {
```
Comment (pxrl) [inline test/SpokePoolClient.SpeedUp.ts:320]: v3 test starts here.

**Comment by pxrl** [comment]
Note: 3 unrelated tests are currently failing in this PR. I've identified the reason for that, and it requires a small change upstream in the contracts repo to mitigate the error. This should be achievable for both existing and new contracts packages.

File: test/utils/utils.ts [hunk 2]
```diff
@@ -346,18 +347,137 @@ export async function buildV2DepositStruct(
...
+  const args = lastEvent?.args;
+  assert.exists(args);
+
+  const { blockNumber, transactionHash, transactionIndex, logIndex } = lastEvent!;
+
+  return {
+    depositId: args!.depositId,
+    originChainId: Number(originChainId),
+    destinationChainId: Number(args!.destinationChainId),
+    depositor: args!.depositor,
```
[the remark this answers is not in the document]
Comment (pxrl) [reply test/utils/utils.ts:413]: Not sure about the linter, but my editor highlights almost everything in the tests directory. Maybe it's being too strict...but there are definite smells down here :(

**Comment by nicholaspai** [comment]
https://github.com/across-protocol/sdk-v2/pull/541#issuecomment-1936468851

File: test/utils/utils.ts [hunk 3]
```diff
@@ -346,18 +347,137 @@ export async function buildV2DepositStruct(
...
+      exclusivityDeadline,
+      message
+    );
+
+  const [events, originChainId] = await Promise.all([
+    spokePool.queryFilter(spokePool.filters.V3FundsDeposited()),
+    spokePool.chainId(),
+  ]);
+
+  const lastEvent = events.at(-1);
```
[the remark this answers is not in the document]
Comment (pxrl) [reply test/utils/utils.ts:368]: Underused imo

**Comment by pxrl** [comment]
> I think this shoud be based on a different PR, see my comment: [#541 (comment)](https://github.com/across-protocol/sdk-v2/pull/541#issuecomment-1936468851)
> 
> However these v3 tests look great!

I've rebased on the updated master and have filled out the remaining tests :+1:

**Comment by pxrl** [comment]
> I wonder if the V3 code should be in a separate file?

For v2/v3 tests affecting the same functionality I'd prefer to keep them in the same file - IMO that's easier to work with.

File: test/utils/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pxrl) [reply test/utils/utils.ts]: Not relevant to this PR anymore because that file isn't changed in the new version :+1:

**Review by nicholaspai — APPROVED** [review]

**Pull request merged**